src/serialization/src/main/kotlin/org/vaccineimpact/api/serialization/ModelBinder.kt

```diff
@@ -3,53 +3,75 @@ package org.vaccineimpact.api.serialization
 import org.vaccineimpact.api.models.validation.CanBeBlank
 import org.vaccineimpact.api.serialization.validation.ValidationException
-import org.vaccineimpact.api.serialization.validation.applyRule
-import org.vaccineimpact.api.serialization.validation.checkMissing
+import org.vaccineimpact.api.serialization.validation.Validator
 import kotlin.reflect.KClass
 import kotlin.reflect.KProperty1
 import kotlin.reflect.full.findAnnotation
 import kotlin.reflect.full.memberProperties
 
-class ModelBinder(private val serializer: Serializer = MontaguSerializer.instance)
+class ModelBinder(private val serializer: Serializer = MontaguSerializer.instance,
+                  private val validator: Validator = Validator(serializer))
 {
     fun <T : Any> deserialize(body: String, klass: Class<T>): T
     {
         val model = serializer.fromJson(preprocessed(body), klass)
         val errors = verify(model)
         if (errors.any())
         {
-            throw ValidationException(errors)
+            throw ValidationException(errors.distinct())
         }
         return model
     }
 
     fun verify(model: Any): List<ErrorInfo>
     {
         val properties = model::class.memberProperties.filterIsInstance<KProperty1<Any, *>>()
-        return properties.flatMap { verify(it, model) }
+        return properties.flatMap { recursiveVerify(it, model) ?: listOf() }
     }
 
-    fun verify(property: KProperty1<Any, *>, model: Any): List<ErrorInfo>
+    private fun recursiveVerify(property: KProperty1<Any, *>, model: Any): List<ErrorInfo>
+    {
+        val errors = verify(property, model)
+
+        // end condition
+        // value is null
+        val value = property.get(model) ?:
+                return errors
+
+        val klass = value::class
+        if (!klass.isData)
+        {
+            // end condition
+            // primitive property
+            return errors
+        }
+
+        val members = klass.memberProperties.filterIsInstance<KProperty1<Any, *>>()
+        return errors + members.flatMap { recursiveVerify(it, value) }
+    }
+
+    private fun verify(property: KProperty1<Any, *>, model: Any): List<ErrorInfo>
     {
         @Suppress("UNCHECKED_CAST")
         val klass = model::class as KClass<Any>
         val errors = mutableListOf<ErrorInfo>()
         val name = serializer.convertFieldName(property.name)
-        val value = property.get(model)
 
-        if (!property.returnType.isMarkedNullable)
+        errors += validator.nullCheck(property.get(model), property, model, name)
+
+        if (property.findAnnotationAnywhere<CanBeBlank>(klass) == null)
         {
-            errors += listOfNotNull(checkMissing(value, name))
+            errors += validator.checkBlank(property, name, model)
         }
-        if (value is String && value.isBlank() && property.findAnnotationAnywhere<CanBeBlank>(klass) == null)
+
+        errors += property.allAnnotations(klass).flatMap { validator.applyRule(it, property, name, model) }
```

Comment:
Why have you switched to passing in the property rather than the value here? Presumably it's a bit slower to keep getting the value by reflection?